idaes/apps/grid_integration/multiperiod/price_taker_model.py

```diff
@@ -0,0 +1,965 @@
+        """
+        # testing if n_integers is valid
+        if n_clusters is not None:
+            if not isinstance(n_clusters, int):
+                raise ValueError(
+                    f"n_clusters must be an integer, but {n_clusters} is not an integer"
+                )
+            if n_clusters < 1:
+                raise ValueError(
+                    f"n_clusters must be > 0, but {n_clusters} is provided."
+                )
+
+        # reconfiguring raw data
+        daily_data = self.generate_daily_data(raw_data)
+
+        # KMeans clustering with the optimal number of clusters
+        kmeans = KMeans(n_clusters=n_clusters).fit(daily_data.transpose())
+        centroids = kmeans.cluster_centers_
+        labels = kmeans.labels_
+
+        # Set any centroid values that are < 1e-4 to 0 to avoid noise
+        centroids = centroids * (centroids >= 1e-4)
+
+        # Compute weight for each cluster by counting its occurrences in the dataset
+        unique_labels, weights_counter = np.unique(labels, return_counts=True)
+
+        # Create dicts for lmp data and the weight of each cluster
+        rep_days_data = {}
+        weights_data = {}
+
+        rep_days_data = pd.DataFrame(
+            centroids.transpose(), columns=range(1, n_clusters + 1)
+        )
+        lmp_data = rep_days_data.to_dict()
+        weights_data = pd.DataFrame(weights_counter)
+        weights_data.index = np.arange(1, len(weights_data) + 1)
+        weights = weights_data.to_dict()
+
+        return lmp_data, weights
+
+    def append_lmp_data(
+        self,
+        file_path,
+        sheet=None,
+        column_name=None,
+        n_clusters=None,
+        horizon_length=None,
+    ):
+        """
+        This function appends LMP data to the PriceTakerModel using single
+        or multiple year signals, as well as using full price data or
+        clustering to use representative days.
+
+        Args:
+            file_path:      path to the file containing LMP data
+            sheet:          if the file is an excel file, the sheet name
+                            where the LMP data is located - (default: None)
+            column_name:    string name of the column to be used for the
+                            LMP signal
+            n_clusters:     number of clusters for the data (representative days)
+                            None if representative days not used - (default: None)
+            horizon_length: if a value is given, this will be used to set the
+                            horizon_length attribute of the PriceTakerModel.
+                            (default: None --> use existing horizon_length (default: 24))
+
+        Returns:
+
+        """
+        if os.path.exists(file_path):
+            path_to_file = file_path
+        else:
+            raise ValueError(
+                f"The file path {file_path} does not exist. Please check your file path."
+            )
+
+        if ".xls" in path_to_file.suffix:
```

Comment:
The extension should be `.xlsx`

Reply:
I agree, but I think this allows for users to read both '.xls' and '.xlsx' files. I can change if we don't want '.xls' extensions to be compatible.

Reply:
Makes sense. No need to change.